### PR TITLE
[LSRA TP]: Skip free heuristics if no free registers are available

### PR DESCRIPTION
**Pull request by kunalspathak**
Noticed that if we don't have free candidates, we would still try all the selection heuristics applicable for selecting "free registers" before we try the ones with "busy registers". Added a condition to by-pass free-registers selection heuristics. Don't expect to see much TP diff, but still good to have.

### CODE REVIEW

**Comment by TIHan** [comment]
TP was improved overall so far.  [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=209100&view=ms.vss-build-web.run-extensions-tab)

**Comment by kunalspathak** [comment]
> TP was improved overall so far. [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=209100&view=ms.vss-build-web.run-extensions-tab)

Seems linux-arm doesn't like this change. I will have to investigate.

image

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/lsra_busy_score.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra_busy_score.h:1]: Why do you need this file? It looks like the data is the same as the "ifdef DEBUG" case in lsra_score.h. Couldn't you just change lsra_score.h to define these using `BUSY_REG_SEL_DEF` and then define that appropriately when including lsra_score.h?

**Review by BruceForstall — COMMENTED** [review]
It's very worrying, in general, to see so much code under `ifdef DEBUG` that appears to be semantically important (as opposed to debug-only asserts or JitDump output). Is that really necessary?

File: src/coreclr/jit/lsra_busy_score.h (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra_busy_score.h:1]: That was exactly what I started with and then did `#define BUSY_REG_SEL_DEF REG_SEL_DEF`, but I was getting compilation errors. Perhaps, there was something that I missed to make it work. I will follow-up with you offline to discuss about the errors.

**Comment by kunalspathak** [comment]
> 

I agree and suggestions welcome. The reason I had to put `freeCandidates == RBM_NONE` check under `#ifdef` is because in both DEBUG/RELEASE we just iterate over all the heuristics and invoke the function lambdas without knowing if they are for free or busy heuristic method. Again, lets chat offline on how we can simplify this.

File: src/coreclr/jit/lsra_busy_score.h (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra_busy_score.h:1]: I missed adding `BUSY_REG_SEL_DEF ` at some call sites and that's what made the compilation errors.

**Comment by kunalspathak** [comment]
@BruceForstall - I didn't change the `DEBUG` code to add extra mapping to see which one needs `freeCandidates` check because I thought it was unnecessary for `DEBUG` purpose. I have removed all the `#ifdef DEBUG` from the heuristic method.

**Review by BruceForstall — APPROVED** [review]

**Comment by kunalspathak** [comment]
Although I see TP regression because of the rearrangement of the generated code, we can clearly see the execution improvement because we would skip multiple checks on the way. Here is the decompiled code where we would jump to `LABEL_218` (on right side).

image
image

**Pull request merged**